File: CS/Form1.cs
using System;
using System.Data;
using DevExpress.Xpo;
using DevExpress.Xpo.DB;
using System.Windows.Forms;
using System.Data.SqlClient;
using System.Collections.Generic;

namespace DynamicDataTable {
    public partial class Form1 : Form {
        private readonly string connectionString = "Data Source=(local);Initial Catalog=Northwind;Integrated Security=True";
        private SqlConnection connection;
        private MSSqlConnectionProvider connectionProvider;
        private Session workSession;
        private Dictionary<string, DynamicDataTableClassInfo> dynamicClasses;
        public Form1() {
            InitializeComponent();
            Init();
        }
        protected void Init() {
            connection = new SqlConnection(connectionString);
            connectionProvider = new MSSqlConnectionProvider(connection, AutoCreateOption.None);
            IDataLayer dal = new SimpleDataLayer(connectionProvider);
            workSession = new Session(dal);
            string[] tablesList = connectionProvider.GetStorageTablesList();
            dynamicClasses = new Dictionary<string, DynamicDataTableClassInfo>(tablesList.Length);
            cmbTables.Properties.Items.AddRange(tablesList);
            cmbTables.SelectedIndexChanged += delegate(object sender, EventArgs args) {
                FillGrid(cmbTables.Text);
            };
        }
        protected void FillGrid(string tableName) {
            grdTableBrowser.DataSource = null;
            gridView1.OptionsBehavior.Editable = false;
            gridView1.Columns.Clear();
            SqlDataAdapter adapter = new SqlDataAdapter(string.Format("SELECT * FROM [{0}]", tableName), connection);
            DataTable dataTable = new DataTable(tableName);
            adapter.FillSchema(dataTable, SchemaType.Source);
            DynamicDataTableClassInfo dataTableClassInfo;
            if (!dynamicClasses.TryGetValue(tableName, out dataTableClassInfo)) {
                dynamicClasses.Add(tableName, new DynamicDataTableClassInfo(workSession.Dictionary, dataTable));
            }
            XPServerCollectionSource dataSource = new XPServerCollectionSource(workSession, dynamicClasses[tableName]);
            grdTableBrowser.DataSource = dataSource;
            //grdTableBrowser.ServerMode = true;
            grdTableBrowser.ForceInitialize();
        }
    }
}

File: VB/Form1.vb
Imports System
Imports System.Data
Imports DevExpress.Xpo
Imports DevExpress.Xpo.DB
Imports System.Windows.Forms
Imports System.Data.SqlClient
Imports System.Collections.Generic

Namespace DynamicDataTable

    Public Partial Class Form1
        Inherits Form

        Private ReadOnly connectionString As String = "Data Source=(local);Initial Catalog=Northwind;Integrated Security=True"

        Private connection As SqlConnection

        Private connectionProvider As MSSqlConnectionProvider

        Private workSession As Session

        Private dynamicClasses As Dictionary(Of String, DynamicDataTableClassInfo)

        Public Sub New()
            InitializeComponent()
            Init()
        End Sub

        Protected Sub Init()
            connection = New SqlConnection(connectionString)
            connectionProvider = New MSSqlConnectionProvider(connection, AutoCreateOption.None)
            Dim dal As IDataLayer = New SimpleDataLayer(connectionProvider)
            workSession = New Session(dal)
            Dim tablesList As String() = connectionProvider.GetStorageTablesList(False)
            dynamicClasses = New Dictionary(Of String, DynamicDataTableClassInfo)(tablesList.Length)
            cmbTables.Properties.Items.AddRange(tablesList)
            AddHandler cmbTables.SelectedIndexChanged, Sub(ByVal sender, ByVal args) FillGrid(cmbTables.Text)
        End Sub

        Protected Sub FillGrid(ByVal tableName As String)
            If TypeOf grdTableBrowser.DataSource Is IDisposable Then
                CType(grdTableBrowser.DataSource, IDisposable).Dispose()
            End If

            grdTableBrowser.DataSource = Nothing
            gridView1.OptionsBehavior.Editable = False
            gridView1.Columns.Clear()
            Dim dataTable As DataTable = New DataTable(tableName)
            Dim adapter As SqlDataAdapter = New SqlDataAdapter(String.Format("SELECT * FROM [{0}]", tableName), connection)
            Try
                adapter.FillSchema(dataTable, SchemaType.Source)
                Dim dataTableClassInfo As DynamicDataTableClassInfo
                If Not dynamicClasses.TryGetValue(tableName, dataTableClassInfo) Then
                    dynamicClasses.Add(tableName, New DynamicDataTableClassInfo(workSession.Dictionary, dataTable))
                End If

                Dim dataSource As XPServerCollectionSource = New XPServerCollectionSource(workSession, dynamicClasses(tableName))
                grdTableBrowser.DataSource = dataSource
                'grdTableBrowser.ServerMode = true;
                grdTableBrowser.ForceInitialize()
            Catch ex As Exception
                Call MessageBox.Show(ex.ToString())
            End Try
        End Sub
    End Class
End Namespace
